include/net/ethernet.h

```diff
@@ -79,12 +79,14 @@ struct ethernet_api {
 	/** Get the device capabilities */
 	enum eth_hw_caps (*get_capabilities)(struct device *dev);
 
+#if defined(CONFIG_NET_VLAN)
 	/** The IP stack will call this function when a VLAN tag is enabled
 	 * or disabled. If enable is set to true, then the VLAN tag was added,
 	 * if it is false then the tag was removed. The driver can utilize
 	 * this information if needed.
 	 */
-	int (*vlan_setup)(struct net_if *iface, u16_t tag, bool enable);
+	int (*vlan_setup)(struct device *dev, u16_t tag, bool enable);
```

Comment:
It's a driver api, not a net_if api, so struct device *dev it is.

Btw, this api is not used by any driver yet, what is the use for it then?

Reply:
It can be used for example by ethernet controller to program the chip for some vlan specific thing. I used it earlier for tracking used vlan's before refactoring the code and the use for it went away. One option is to remove it if we really have no use for it atm.

I really think we need to use iface here instead of dev. It is impossible to figure out what network interface is used if dev is given as a parameter here.

Reply:
Than let's add an iface parameter along with dev. iface would be the vlan iface

Reply:
Then there is really no need to add dev parameter as that can be figured out from iface.

Reply:
As I said, ethernet_api is a device driver api, not a network interface api. so dev should be present. Unless we follow any rules...

Reply:
I do not see point to have a parameter in an api that can be deducted from the other parameter. But if you really insists, then sure add the dev param there. But we definitely need the iface parameter too.

Reply:
It's just the point of being consistent (I think we care about that), plus you'll need to access the device's context maybe in the implementation in driver side. 4 parameters is also fine from stack usage point of view. I'll add it.